<a href="https://colab.research.google.com/github/surajk150741/Data_Science_Projects/blob/main/Spell%20Correction/TextBlob.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Spell correction

- To compare two given text documents for their similarity ratio.
- To apply Textblob on Error_Version for corrected version and again calculating the similarity with the Error_Version.

**Importing Data**

In [ ]:

from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
t1 = open('/content/drive/MyDrive/Error_Version.txt','r')

In [ ]:
t2 = open('/content/drive/MyDrive/Truth_Version.txt','r')

In [ ]:
t1.name

'/content/drive/MyDrive/Error_Version.txt'

In [ ]:
t2.name

'/content/drive/MyDrive/Truth_Version.txt'

In [ ]:
t1.mode

'r'

In [ ]:
t2.mode

'r'

In [ ]:

FileContent1 = t1.read()
print(FileContent1)

the students let us start the first lecture of this course this is basically a anline course so therefore you need to be very clear about what you are going to learn and how yeu are going to be evaluated first of all this is a course on phase transformation in materials and i will be teaching all the lectures and it is a thirty lecture corse that means there will be sixte twenty five minutes lecture and every week w will be exposed to i am not sure now about four to six lectures so that means about two to three hours of lectures so myself is persummiswas and i am a faculty member at the department of material cense engineering opiitianpull so thuts the detals about me my email address is available on the screen you can dialaly write to me whenever you have any serious problems and in addition to that to handle the courseexams and also grading of your answer seets although i know this will be all done all line will be taken care by the teaching assistems so two my students they will be

In [ ]:

FileContent2 = t2.read()
print(FileContent2)

dear students let us start the first lecture of this course this is basically a online
course so therefore you need to be very clear about what you are going to learn and how
you are going to be evaluated first of all this is a course on phase transformation in
materials and i will be teaching all the lectures and it is a thirty lecture course that means
there will be sixty twenty five minutes lecture and every week will be i suppose to i am not
sure now about four to six lectures so that means about two to three hours of lectures
so myself is professor biswas and i am a faculty member at the department of material science
engineering of iit kanpur so thats the details about me my email address is available on
the screen you can daily write to me whenever you have any serious problems and in addition
to that to handle the course exams and also grading appear answer sheets i know this will
be done online will be taken care by the teaching assistance
so two of my students they will be h

In [ ]:
str(FileContent2)

'\ufeffdear students let us start the first lecture of this course this is basically a online\ncourse so therefore you need to be very clear about what you are going to learn and how\nyou are going to be evaluated first of all this is a course on phase transformation in\nmaterials and i will be teaching all the lectures and it is a thirty lecture course that means\nthere will be sixty twenty five minutes lecture and every week will be i suppose to i am not\nsure now about four to six lectures so that means about two to three hours of lectures\nso myself is professor biswas and i am a faculty member at the department of material science\nengineering of iit kanpur so thats the details about me my email address is available on\nthe screen you can daily write to me whenever you have any serious problems and in addition\nto that to handle the course exams and also grading appear answer sheets i know this will\nbe done online will be taken care by the teaching assistance\nso two of my studen

## Using difflib

In [ ]:
import difflib

In [ ]:
f1 = FileContent1.split()
f2 = FileContent2.split()

In [ ]:
seq_mat = difflib.SequenceMatcher()

seq_mat.set_seqs(f1, f2)
print("Similarity Ratio            : {}".format(seq_mat.ratio()))

Similarity Ratio            : 0.7400474558397048


## Now, I am gonna use nltk for comparing two documents

In [ ]:
!pip install nltk

In [ ]:
import nltk

In [ ]:
nltk.download('all')

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/abc.zip.
[nltk_data]    | Downloading package alpino to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/alpino.zip.
[nltk_data]    | Downloading package biocreative_ppi to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping corpora/biocreative_ppi.zip.
[nltk_data]    | Downloading package brown to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/brown.zip.
[nltk_data]    | Downloading package brown_tei to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/brown_tei.zip.
[nltk_data]    | Downloading package cess_cat to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/cess_cat.zip.
[nltk_data]    | Downloading package cess_esp to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/cess_esp.zip.
[nltk_data]    | Downloading package chat80 to /root/nltk_data...
[nltk_data]    |   Unzipp

True

**Here, I am going to calculate the similarity ratio between Error_Version and Truth Version**

In [ ]:
# Program to measure the similarity between 
# two sentences using cosine similarity.
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
  
# X = input("Enter first string: ").lower()
# Y = input("Enter second string: ").lower()
X =FileContent1
Y =FileContent2
  
# tokenization
X_list = word_tokenize(X) 
Y_list = word_tokenize(Y)
  
# sw contains the list of stopwords
sw = stopwords.words('english') 
l1 =[];l2 =[]
  
# remove stop words from the string
X_set = {w for w in X_list if not w in sw} 
Y_set = {w for w in Y_list if not w in sw}
  
# form a set containing keywords of both strings 
rvector = X_set.union(Y_set) 
for w in rvector:
    if w in X_set: l1.append(1) # create a vector
    else: l1.append(0)
    if w in Y_set: l2.append(1)
    else: l2.append(0)
c = 0
  
# cosine formula 
for i in range(len(rvector)):
        c+= l1[i]*l2[i]
cosine = c / float((sum(l1)*sum(l2))**0.5)
print("similarity: ", cosine)

similarity:  0.5999798083293154


**- I found that the similarity ration is 0.5999798 and percentage would be 59.99798%.**

## Now, I am gonna perform spelling correction with the help of TextBlob.

In [ ]:
!pip install textblob

In [ ]:
from textblob import TextBlob
import re

In [ ]:
import re
str=re.findall("[a-zA-Z,.]+",FileContent1)
updated_docx=(" ".join(str))
print(updated_docx)

the students let us start the first lecture of this course this is basically a anline course so therefore you need to be very clear about what you are going to learn and how yeu are going to be evaluated first of all this is a course on phase transformation in materials and i will be teaching all the lectures and it is a thirty lecture corse that means there will be sixte twenty five minutes lecture and every week w will be exposed to i am not sure now about four to six lectures so that means about two to three hours of lectures so myself is persummiswas and i am a faculty member at the department of material cense engineering opiitianpull so thuts the detals about me my email address is available on the screen you can dialaly write to me whenever you have any serious problems and in addition to that to handle the courseexams and also grading of your answer seets although i know this will be all done all line will be taken care by the teaching assistems so two my students they will be 

In [ ]:
new_doc = TextBlob(updated_docx)

FileContent3 = new_doc.correct()
print(FileContent3)



the students let us start the first lecture of this course this is basically a online course so therefore you need to be very clear about what you are going to learn and how you are going to be evacuated first of all this is a course on phase transformation in materials and i will be teaching all the lectures and it is a thirty lecture course that means there will be sixth twenty five minutes lecture and every week w will be exposed to i am not sure now about four to six lectures so that means about two to three hours of lectures so myself is persummiswas and i am a faculty member at the department of material sense engineering opiitianpull so thus the details about me my email address is available on the screen you can dilly write to me whenever you have any serious problems and in addition to that to handle the courseexams and also trading of your answer seems although i know this will be all done all line will be taken care by the teaching assisted so two my students they will be ha

In [ ]:
s = ""
for i in range(len(FileContent3)):
  s += FileContent3[i]


In [ ]:
f3 = s.split()

In [ ]:
import difflib

In [ ]:
seq_mat = difflib.SequenceMatcher()

seq_mat.set_seqs(f2, f3)
print("Similarity Ratio            : {}".format(seq_mat.ratio()))

Similarity Ratio            : 0.741795175958877


In [ ]:
listt = []

In [ ]:
si = difflib.SequenceMatcher(None, s, FileContent2)
for tag, i1, i2, j1, j2 in si.get_opcodes():
  listt.append(tag)
  print('{:7}   s[{}:{}] --> FileContent2[{}:{}] {!r:>8} --> {!r}'.format(
  tag, i1, i2, j1, j2, s[i1:i2], FileContent2[j1:j2]))

replace   s[0:86] --> FileContent2[0:88] 'the students let us start the first lecture of this course this is basically a online ' --> '\ufeffdear students let us start the first lecture of this course this is basically a online\n'
equal     s[86:173] --> FileContent2[88:175] 'course so therefore you need to be very clear about what you are going to learn and how' --> 'course so therefore you need to be very clear about what you are going to learn and how'
replace   s[173:174] --> FileContent2[175:176]      ' ' --> '\n'
equal     s[174:197] --> FileContent2[176:199] 'you are going to be eva' --> 'you are going to be eva'
replace   s[197:356] --> FileContent2[199:358] 'cuated first of all this is a course on phase transformation in materials and i will be teaching all the lectures and it is a thirty lecture course that means ' --> 'luated first of all this is a course on phase transformation in\nmaterials and i will be teaching all the lectures and it is a thirty lecture course that mean

In [ ]:
def CountFrequency(my_list):
 
    # Creating an empty dictionary
    freq = {}
    for item in my_list:
        if (item in freq):
            freq[item] += 1
        else:
            freq[item] = 1
 
    for key, value in freq.items():
        print ("% s : % d"%(key, value))
 


In [ ]:
CountFrequency(listt)

replace :  68
equal :  67


- Converting object type of FileContent3 from "TextBlob"  to "str".

In [ ]:
type(s)

str

# Applying enchant to correct only those words, which are in dictionary

In [ ]:
!apt install -qq enchant
!pip install pyenchant

The following package was automatically installed and is no longer required:
  libnvidia-common-460
Use 'apt autoremove' to remove it.
The following additional packages will be installed:
  aspell aspell-en dictionaries-common emacsen-common hunspell-en-us
  libaspell15 libenchant1c2a libhunspell-1.6-0 libtext-iconv-perl
Suggested packages:
  aspell-doc spellutils wordlist hunspell openoffice.org-hunspell
  | openoffice.org-core libenchant-voikko
The following NEW packages will be installed:
  aspell aspell-en dictionaries-common emacsen-common enchant hunspell-en-us
  libaspell15 libenchant1c2a libhunspell-1.6-0 libtext-iconv-perl
0 upgraded, 10 newly installed, 0 to remove and 40 not upgraded.
Need to get 1,312 kB of archives.
After this operation, 5,353 kB of additional disk space will be used.
Preconfiguring packages ...
Selecting previously unselected package libtext-iconv-perl.
(Reading database ... 148489 files and directories currently installed.)
Preparing to unpack .../0-libt

In [ ]:
import enchant

In [ ]:
d = enchant.Dict("en_US")
  
type(d)

enchant.Dict

In [ ]:
test_string = ""

In [ ]:
print("Hi Suraj")
for i in range(len(f1)):
    if d.check(f1[i]):
        test_string += f1[i]
        test_string += " "
    
    else:
        new = TextBlob(f1[i])
        result = new.correct()
        for j in range(len(result)):
            
            test_string += result[j]
        test_string += " "
      
test_string

Hi Suraj


"\ufeffthe students let us start the first lecture of this course this is basically a online course so therefore you need to be very clear about what you are going to learn and how you are going to be evaluated first of all this is a course on phase transformation in materials and i will be teaching all the lectures and it is a thirty lecture course that means there will be sixth twenty five minutes lecture and every week w will be exposed to i am not sure now about four to six lectures so that means about two to three hours of lectures so myself is persummiswas and i am a faculty member at the department of material sense engineering opiitianpull so thus the details about me my email address is available on the screen you can dilly write to me whenever you have any serious problems and in addition to that to handle the courseexams and also grading of your answer seems although i know this will be all done all line will be taken care by the teaching assisted so two my students they wil

In [ ]:
print(test_string)

the students let us start the first lecture of this course this is basically a online course so therefore you need to be very clear about what you are going to learn and how you are going to be evaluated first of all this is a course on phase transformation in materials and i will be teaching all the lectures and it is a thirty lecture course that means there will be sixth twenty five minutes lecture and every week w will be exposed to i am not sure now about four to six lectures so that means about two to three hours of lectures so myself is persummiswas and i am a faculty member at the department of material sense engineering opiitianpull so thus the details about me my email address is available on the screen you can dilly write to me whenever you have any serious problems and in addition to that to handle the courseexams and also grading of your answer seems although i know this will be all done all line will be taken care by the teaching assisted so two my students they will be h

In [ ]:
listtt = []

In [ ]:
sii = difflib.SequenceMatcher(None, test_string, FileContent2)
for tag, i1, i2, j1, j2 in sii.get_opcodes():
  listtt.append(tag)
  print('{:7}   test_string[{}:{}] --> FileContent2[{}:{}] {!r:>8} --> {!r}'.format(
  tag, i1, i2, j1, j2, test_string[i1:i2], FileContent2[j1:j2]))

equal     test_string[0:1] --> FileContent2[0:1] '\ufeff' --> '\ufeff'
replace   test_string[1:87] --> FileContent2[1:88] 'the students let us start the first lecture of this course this is basically a online ' --> 'dear students let us start the first lecture of this course this is basically a online\n'
equal     test_string[87:174] --> FileContent2[88:175] 'course so therefore you need to be very clear about what you are going to learn and how' --> 'course so therefore you need to be very clear about what you are going to learn and how'
replace   test_string[174:175] --> FileContent2[175:176]      ' ' --> '\n'
equal     test_string[175:261] --> FileContent2[176:262] 'you are going to be evaluated first of all this is a course on phase transformation in' --> 'you are going to be evaluated first of all this is a course on phase transformation in'
replace   test_string[261:357] --> FileContent2[262:358] ' materials and i will be teaching all the lectures and it is a thirty lecture cours

In [ ]:
def CountFrequency(my_list):
 
    # Creating an empty dictionary
    freq = {}
    for item in my_list:
        if (item in freq):
            freq[item] += 1
        else:
            freq[item] = 1
 
    for key, value in freq.items():
        print ("% s : % d"%(key, value))
 


In [ ]:
CountFrequency(listtt)

equal :  68
replace :  68
